feat(iota-genesis-builder): Add configurable delegator to genesis CLI

nonast opened the pull request:
# Description of change

To make the genesis delegator(s) configurable by the cli, we need to make changes to two iota commands:
1. `iota genesis`
2. `iota genesis-ceremony` 

This PR contains the progress of this task, in which the changes to both commands have been made to allow for a custom delegation configuration through the CLI. The `iota-genesis` command expects a single delegator when a migration snapshot has been provided. The `iota genesis-ceremony` now contains a new sub command called `init-delegations` and expects a path to a `.csv` file as argument, which contains the custom delegation configuration. The command is expected to be run before the `build-unsigned-checkpoint` command.

This csv file has the following format: 

```
delegator,validator,amount-nanos-to-stake,amount-nanos-to-pay-gas
<delegator1-address>,<validator-1-address>,2000000000000000,5000000000
<delegator1-address>,<validator-2-address>,3000000000000000,5000000000
<delegator2-address>,<validator-3-address>,4500000000000000,5000000000
```


## Links to any relevant issues

fixes #4218 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Tested the local genesis:
1.  `cargo run --release --bin iota genesis --working-dir ./test_iota_config -f --with-faucet --num-validators 2 --remote-migration-snapshots https://stardust-objects.s3.eu-central-1.amazonaws.com/iota/alphanet/test/stardust_object_snapshot.bin.gz --delegator 0x4f72f788cdf4bb478cf9809e878e6163d5b351c82c11f1ea28750430752e7892`
2. `cargo run --release --bin iota start --network.config test_iota_config --with-faucet`

Tested the following affected example:
- `cargo run --example snapshot_only_test_outputs --features="test-outputs" iota --snapshot-path ../latest-full_snapshot.bin --delegator 0x4f72f788cdf4bb478cf9809e878e6163d5b351c82c11f1ea28750430752e7892`

Tested running the genesis ceremony, which resulted in a `genesis.blob` and `migration.blob`, which were then used to run a local network with `iota start`. Using the explorer on this local network showed the correct delegations.

todo: the `validate-state` command of the genesis ceremony is broken when used with a migrated state. This step was therefore skipped in the tests mentioned above.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

nonast committed:
feat(iota-genesis-builder): add delegator as cli arg for genesis

nonast committed:
add bail when delegator missing

nonast committed:
fix genesis tests

nonast committed:
feat(genesis): add delegator map type to genesis builder

miker83z committed:
refactor(iota-genesis-builder): add comments and restructure resolve_token_distribution_schedule

miker83z commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -118,6 +114,8 @@ pub struct Builder {
     genesis_stake: GenesisStake,
     migration_sources: Vec<SnapshotSource>,
     migration_tx_data: Option<MigrationTxData>,
+    delegator: Option<String>,
```
This can definitively be of type `IotaAddress`

nonast replied:
done

miker83z commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -907,6 +960,8 @@ impl Builder {
                 .save(file)?;
         }
 
+        // todo: probably need to save the delegator and delegator_map?
```
Yes! 

For the delegator it can be something like the `serde_json::to_string(&self.migration_sources)` above. 

For the map you can use the `to_csv`method.

nonast replied:
done

miker83z commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -820,6 +871,8 @@ impl Builder {
             genesis_stake: Default::default(),
             migration_sources,
             migration_tx_data,
+            delegator: None,     // todo: Probably need to load the delegator?
+            delegator_map: None, // todo: Probably need to load the delegator_map?
```
Yes, check the comment to the `save()`  below. For delegator map you find the `from_csv`.

nonast replied:
done

miker83z commented on crates/iota/src/genesis_ceremony.rs:
```diff
@@ -118,6 +118,12 @@ pub enum CeremonyCommand {
         #[clap(long, name = "iota|<full-url>", help = "Remote migration snapshots.")]
         #[arg(num_args(0..))]
         remote_migration_snapshots: Vec<SnapshotUrl>,
+        #[clap(
+            long,
+            help = "Path to the delegator map file.",
+            name = "delegator_map.csv"
+        )]
+        delegator_map: Option<PathBuf>,
```
Can you create a dedicated command similar to `InitTokenDistributionSchedule`?

nonast replied:
done: `InitDelegatorMap`

miker83z commented on crates/iota/src/iota_commands.rs:
```diff
@@ -274,6 +274,8 @@ pub enum IotaCommand {
         #[clap(long, name = "iota|<full-url>")]
         #[arg(num_args(0..))]
         remote_migration_snapshots: Vec<SnapshotUrl>,
+        #[clap(long, help = "Specify the delegator address")]
+        delegator: Option<String>,
```
I think `IotaAddress` can be used directly here.

nonast replied:
Currently, the delegator is supplied as stardust address. Are we planning to supply it as IotaAddress (0x..) for mainnet?

miker83z replied:
Yes the delegator will come from the Destination Address of the Address Swap Map #4314.

nonast replied:
ok, done

nonast committed:
refactor(iota-genesis-builder): genesis cli arg to IotaAddress and add load/save DelegatorMap

nonast committed:
feat(genesis-ceremony): add InitDelegatorMap command

kodemartin commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,110 @@ impl TokenDistributionScheduleBuilder {
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct DelegatorDistribution {
```
Nit: This rather implies that the struct is about a distribution of delegators. Simply naming it `Delegation` would be more descriptive and accurate I reckon.

nonast replied:
renamed

kodemartin commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,110 @@ impl TokenDistributionScheduleBuilder {
+pub struct DelegatorMap {
+    // Maps a delegator address to a tuple containing the address of a validator (1nd element of
+    // the tuple) that will receive an amount of nanos to stake (2nd element) and an amount as gas
+    // payment (3rd element).
```
Map is not to a `tuple` anymore. You might consider updating the doc. I suggest using `\\\` type comments as well.

nonast replied:
Rewritten

kodemartin commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,110 @@ impl TokenDistributionScheduleBuilder {
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct DelegatorMap {
```
Nit: Following the previous comments this could be renamed to `Delegations` .

nonast replied:
Renamed

kodemartin commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,110 @@ impl TokenDistributionScheduleBuilder {
+                    });
+                allocations
+            },
+        );
```
Nit: Using `fold` seems like an overkill since we only have a single `delegator`. Might consider just inserting `validators.into_iter().map(|address| ValidatorAllocation {} ).collect()` into the hashmap for that `delegator` instead.

nonast replied:
Simplified as suggested

kodemartin commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,110 @@ impl TokenDistributionScheduleBuilder {
+    pub address: IotaAddress,
+    // The amount of nanos to stake to the validator
+    pub amount_nanos_to_stake: u64,
+    /// The amount of nanos to transfer as gas payment to the validator
+    pub amount_nanos_to_pay_gas: u64,
```
Is this used?

miker83z replied:
Additional logic will be added to burn gas tokens from a delegator and allocate these to a validator (not as stake, but as payment).

kodemartin commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -118,6 +116,8 @@ pub struct Builder {
     migration_sources: Vec<SnapshotSource>,
     migration_tx_data: Option<MigrationTxData>,
+    delegator: Option<IotaAddress>,
+    delegator_map: Option<DelegatorMap>,
```
I would suggest wrapping both fields into a single one using a 
```rust
enum GenesisDelegation {
       OneToAll(IotaAddress),
       ManyToMany(DelegatorMap),
}
```

This will make the API more transparent and type-safe.

nonast replied:
Great suggestion, didn't think of this. Added this new enum.

kodemartin commented on crates/iota/src/iota_commands.rs:
```diff
@@ -988,6 +994,16 @@ async fn genesis(
+            // Add a delegator account to the genesis.
+            genesis_conf = genesis_conf.add_delegator(delegator);
+        } else {
+            bail!("A delegator must be supplied when migration snapshots are provided.");
```
Typically we use lower-case first letters in error messages. Please review other error messages in the patch as well.
```suggestion
            bail!("a delegator must be supplied when migration snapshots are provided.");
```

nonast replied:
Done

kodemartin commented on crates/iota/src/iota_commands.rs:
```diff
@@ -274,6 +274,8 @@ pub enum IotaCommand {
         #[arg(num_args(0..))]
         remote_migration_snapshots: Vec<SnapshotUrl>,
+        #[clap(long, help = "Specify the delegator address")]
+        delegator: Option<IotaAddress>,
```
This surely breaks

https://github.com/iotaledger/iota/blob/6dab9e57e28af5ae6648cf5728326ff33450e9fe/docker/pg-services-local/docker-compose.yaml#L18-L25

where we start the local network with test migration data. Could you please update the command by adding the previously hardcoded address?

kodemartin replied:
There are two options:

1. Either add a default value for `delegator` here, or
2. Add the `IotaAddress` values in `docker-compose.yml`

miker83z replied:
The "default" delegator is such only for that `alphanet/test/stardust_object_snapshot.bin.gz` snapshot. Other than that, given #4314, we will have no predefined delegator. So, @nonast, here it is safe to hardcode the original default delegator address (though, as long as the snapshot URL returns an unaltered version).

nonast replied:
Added the address as hardcoded argument

kodemartin replied:
Just tested it. You need to add the argument to the `Start` subcommand as well...

nonast replied:
Yes you're right.. I thought this part of the code was still the `genesis` command somehow.. Fixed it now.

kodemartin commented on crates/iota-genesis-builder/src/lib.rs:
```diff
@@ -253,75 +265,116 @@ impl Builder {
-                // Genesis stake is already included
-                schedule
+
+        if let Some(schedule_without_migration) = self.token_distribution_schedule.take() {
+            if !is_genesis_with_migration || schedule_without_migration.contains_timelocked_stake()
+            {
+                // Case 1.1 and 2.1.1
+                schedule_without_migration
```
As you suggest also in the documentation case 2.1.1 includes the migration so `schedule_without_migration` is rather misleading a name. So I suggest reverting back to `schedule`.

nonast replied:
Correct, changed

kodemartin reviewed (commented):
@nonast looks good in general. Made a few minor suggestions that you could consider.

nonast committed:
refactor(iota-genesis-builder): renaming structs and replace fold() usage

nonast committed:
feat(iota-genesis-builder): add GenesisDelegation enum

nonast committed:
chore(iota): correct bail messages to start with lower case

nonast committed:
refactor(iota-genesis-builder): rename schedule_without_migration to schedule for clarity

miker83z commented on crates/iota-config/src/genesis.rs:
```diff
@@ -64,13 +64,13 @@ impl PartialEq for Genesis {
             && {
-                let this = self.checkpoint.auth_sig();
-                let other = other.checkpoint.auth_sig();
+            let this = self.checkpoint.auth_sig();
```
I think you formatter is failing here

nonast replied:
Gotcha, missed the formatting

miker83z committed:
feat(iota-genesis-builder): create a token allocation to pay validator with gas

miker83z committed:
feat(iota-genesis-builder): optimize timelock and gas objects picks

miker83z committed:
fix format

nonast committed:
fix: add hardcoded delegator arg to docker config

nonast committed:
fix(iota-genesis-builder): fix csv read and write functions

nonast committed:
chore: fix clippy warnings

miker83z committed:
fix pick_objects_for_allocation logic

valeriyr reviewed (approved):
Looks good and well documented!
There are several pretty complex functions, the correctness of which is difficult to check visually; I hope we have tests for all branches of logic.

miker83z committed:
Merge branch 'develop' into sc-platform/custom-genesis-delegator

DaughterOfMars commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,114 @@ impl TokenDistributionScheduleBuilder {
+pub struct ValidatorAllocation {
+    /// The validator address receiving the stake and/or gas payment
+    pub validator: IotaAddress,
+    // The amount of nanos to stake to the validator
```
```suggestion
    /// The amount of nanos to stake to the validator
```

DaughterOfMars committed:
minor improvements

DaughterOfMars commented:
Pushed some minor changes, feel free to revert if you don't like them

nonast commented:
> Pushed some minor changes, feel free to revert if you don't like them

Thank you for the changes! I reverted back one part: the serialization of `ValidatorAllocation` in `to_csv()`, due to serialization of maps not being supported in the csv crate. This forced me to write it out like this.

nonast committed:
fix: add delegator to start cmd and partially revert minor improvement

kodemartin reviewed (approved):
lgtm :bird:

Dkwcs commented on crates/iota-config/src/genesis.rs:
```diff
@@ -568,7 +563,17 @@ impl TokenDistributionSchedule {
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(rename_all = "kebab-case")]
 pub struct TokenAllocation {
```
Maybe we can have minor description here of what is the purpose of this struct?

Dkwcs commented on crates/iota-config/src/genesis.rs:
```diff
@@ -628,3 +633,113 @@ impl TokenDistributionScheduleBuilder {
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct ValidatorAllocation {
```
Maybe we can have minor description here of what is the purpose of this struct?

nonast replied:
Done

Dkwcs commented on crates/iota-genesis-builder/examples/snapshot_only_test_outputs.rs:
```diff
@@ -32,6 +29,8 @@ enum Snapshot {
         #[clap(long, help = "Path to the Iota Hornet full-snapshot file")]
         snapshot_path: String,
+        #[clap(long, help = "Specify the delegator address")]
+        delegator: String,
```
Assume we can set it as IotaAddress, but it’s really minor(because it’s an example).

nonast replied:
Yes, you're right. I changed it to IotaAddress, which simplifies its usage a bit in the main function. 👍

DaughterOfMars commented:
> > Pushed some minor changes, feel free to revert if you don't like them
> 
> Thank you for the changes! I reverted back one part: the serialization of `ValidatorAllocation` in `to_csv()`, due to serialization of maps not being supported in the csv crate. This forced me to write it out like this.

I'm confused. I was serializing a struct which should be supported according to their docs.

nonast committed:
refactor: add comments and adjust example argument

nonast commented:
> > > Pushed some minor changes, feel free to revert if you don't like them
> > 
> > 
> > Thank you for the changes! I reverted back one part: the serialization of `ValidatorAllocation` in `to_csv()`, due to serialization of maps not being supported in the csv crate. This forced me to write it out like this.
> 
> I'm confused. I was serializing a struct which should be supported according to their docs.



> > > Pushed some minor changes, feel free to revert if you don't like them
> > 
> > 
> > Thank you for the changes! I reverted back one part: the serialization of `ValidatorAllocation` in `to_csv()`, due to serialization of maps not being supported in the csv crate. This forced me to write it out like this.
> 
> I'm confused. I was serializing a struct which should be supported according to their docs.

_As discussed on Slack:_ 
- Due to the flatten attribute, the rust crate fails to serialize the struct. This is apparently a known bug: https://github.com/BurntSushi/rust-csv/issues/188
- Agreed to leave as is.

miker83z committed:
fix fmt

nonast committed:
fix clippy issue

nonast committed:
Merge branch 'develop' into sc-platform/custom-genesis-delegator

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+        // Then, try to create new token allocations for each validator using the
+        // objects fetched above
+        for validator_allocation in validators_allocations {
+            // The validaotr address
```
```suggestion
            // The validator address
```

nonast replied:
fixed

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+            let mut timelock_objects =
+                pick_objects_for_allocation(timelocks_pool, target_stake, &mut timelock_surplus);
+            if !timelock_objects.to_burn.is_empty() {
+                // Inside this block, the previous surplus is empty so we update it (possibly)
```
That's a bit confusing to me, why is it empty here? Is it because `pick_objects_for_allocation` enforces that? If yes, maybe we should add that to the comment. At first look I would say that the if condition checks that it's not empty, but then that's the `to_burn` object.

muXxer replied:
Btw, shouldn't we check that it is really empty before replacing it?

muXxer replied:
Another thing that comes to my mind.... why do we even "burn" something? Is it because we consume those "UTXO's"? If yes, shouldn't we rather call it `to_consume` instead of `to_burn`?
Burning, at least in my head, is only used if we actually destroy tokens that are gone forever afterwards.

nonast replied:
We are technically "draining" the surplus object from its surplus_nanos. Therefore, the surplus object can be burned. 

In the `pick_objects_for_allocation` function, we try and take all the nanos that we need from the surplus object first. If we have "drained" the surplus object, we place it in the `timelock_objects.to_burn` Vec. This means that the previous surplus is empty, and it needs to be replaced with a new one (picked in `pick_objects_for_allocation` because the previous allocation might have needed some extra nanos). 

Inside pick_objects_for_allocation this happens:
surplus = Coin(50)
1. if target_stake == 49 then  surplus = Coin(1); timelock_objects.amount_nanos = 49; [timelock_objects.to](http://timelock_objects.to/)_burn = empty
2. if target_stake == 50 then  surplus = empty; timelock_objects.amount_nanos = 50; [timelock_objects.to](http://timelock_objects.to/)_burn = [surplus]
3. if target_stake == 51 then  surplus = empty; timelock_objects.amount_nanos = 51; [timelock_objects.to](http://timelock_objects.to/)_burn = [surplus, timelock_1]

> Btw, shouldn't we check that it is really empty before replacing it?

This is a given, since we're in the scope of `timelock_objects.to_burn.is_empty()`. If the `to_burn` field is empty, it means we still have nanos in the surplus object. Would you advice adding a check here out of safety/clarity? Or could some comments explaining this be sufficient?

> Another thing that comes to my mind.... why do we even "burn" something? Is it because we consume those "UTXO's"? If yes, shouldn't we rather call it to_consume instead of to_burn?
Burning, at least in my head, is only used if we actually destroy tokens that are gone forever afterwards.

Indeed, at first glance I also think we could use another term for this, something more in line with what's being used throughout the code: to `destroy` empty objects. 
However, these objects are not really "consumed" in the actual object database yet. This happens in a later stage in the `build_unsigned_genesis_data` phase of the ceremony. First, the required genesis objects are created. After this, the genesis objects are being "cleansed" from the to be burned objects that we have specified earlier.

miker83z replied:
Changed burn into destroy here [1ef83d2](https://github.com/iotaledger/iota/pull/4346/commits/1ef83d2ed49087ab6f4e545de2316de1cf59229b)

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+            self.timelocks_to_split
+                .push((surplus_timelock, surplus_nanos, delegator));
+        }
+        // In the case a the gas coin, it must be burned and the surplus re-allocated to
```
Is there a word missing? I don't get the beginning of the sentence.

nonast replied:
fixed

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+    /// The surplus amount for that coin object.
+    surplus_nanos: u64,
+    /// Possibly indicate a timelock stake expiration.
+    timestamp: u64,
```
Wouldn't it be better to use an option here?
I mean someone could still create an UTXO with a timelock and set the timelock to zero.
Maybe that would lead to the same result, but I'm just wondering if using `>0` logic is really the cleanest way.

nonast replied:
Hmm.. I guess it makes sense to use an Option here if there should be a clear distinction between surplus being 0 and non-existent. In this case, I don't think it's necessary to convert this to an Option. It would add additional complexity to the code (checks).

miker83z replied:
During the migration from UTXO to Object, [here](https://github.com/iotaledger/iota/blob/c42f61419852475adfbeb4eea8ab6aecf12f43aa/crates/iota-types/src/timelock/timelock.rs#L64) we check if the timelock still holds at the "target milestone timestamp". If it is [not timelocked at that time](https://github.com/iotaledger/iota/blob/c42f61419852475adfbeb4eea8ab6aecf12f43aa/crates/iota-genesis-builder/src/stardust/migration/migration.rs#L215) then we treat is as a simple basic output. This process is performed before this genesis building step, so we can't have unwanted behaviors like your scenario.

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+    }
+
+    // Destroy the `CoinSurplus` and take the fields.
+    pub fn take(self) -> (Option<ObjectRef>, u64) {
```
And what about the timestamp here? Is that logic not necessary in case we take?

nonast replied:
Not necessary in our use-case, where we only need those two fields to be added to the `timelocks_to_split` and `gas_coins_to_burn` Vecs. This is the only situation where `take` is used.

nonast committed:
chore: fix typos in comments

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+            // The validator address
+            let validator = validator_allocation.validator;
+            // The target amount of nanos to be staked, either with timelock or gas objects
+            let mut target_stake = validator_allocation.amount_nanos_to_stake;
```
we should also call that `target_stake_nanos`

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+            // The target amount of nanos to be staked, either with timelock or gas objects
+            let mut target_stake = validator_allocation.amount_nanos_to_stake;
+            // The gas to pay to the validator
+            let gas_to_pay = validator_allocation.amount_nanos_to_pay_gas;
```
we should also call that `gas_to_pay_nanos`

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
 ) -> AllocationObjects {
-    let mut amount_nanos = 0;
-    let mut surplus_nanos = 0;
+    let mut allocation_tot_amount = 0;
```
```suggestion
    let mut allocation_tot_amount_nanos = 0;
```

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
-    let gas_coins_pool = migration_objects.get_gas_coins_by_owner(delegator);
-    if timelocks_pool.is_none() && gas_coins_pool.is_none() {
-        anyhow::bail!("no timelocks or gas-coin objects found for delegator {delegator:?}");
+    if allocation_tot_amount < target_amount {
```
Please put a comment related to reentrance of the function.

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+                            None
+                        }
+                    } else {
+                        // Break
```
```suggestion
                        // Break the map_while
```

muXxer commented on crates/iota-genesis-builder/src/stake.rs:
```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+                                staked_with_timelock.push((difference_from_target, timestamp));
+                            }
+                            allocation_tot_amount += difference_from_target;
+                            // Do NOT place `obj_ref` in `to_burn` and break
```
```suggestion
                            // Do NOT place `obj_ref` in `to_burn` because it is reused in the CoinSurplus and then break the map_while
```

muXxer reviewed (approved):
LGTM, Just some small renaming and comment fixes

miker83z committed:
fix(iota-genesis-builder): comments and names to make it clearer

miker83z committed:
refactor(iota-genesis-builder): use the destroy verb instead of burn

miker83z committed:
Merge branch 'develop' into sc-platform/custom-genesis-delegator

miker83z committed:
fix(iota-genesis-builder): update genesis_config_snapshot_matches snapshot

miker83z committed:
Merge branch 'develop' into sc-platform/custom-genesis-delegator

miker83z committed:
fix(iota-e2e-tests): update test_full_node_load_migration_data with delegator
